src/satisfia/agents/makeMDPAgentSatisfia.py

```diff
@@ -0,0 +1,250 @@
+		if verbose or debug:
+			print("minAdmissibleQ, state", state, "action", action, "...")
+
+		# register (state, action) in global store (could be anywhere, but here is just as fine as anywhere else)
+		stateActionPairsSet.insert((state, action))
+
+		Edel = expectedDelta(state, action)
+		if state.terminateAfterAction:
+			q = Edel
+		else:
+			# Bellman equation
+			q = distribution.infer(lambda: self.minAdmissibleV(self.world.transition(state, action))).E() # recursion
+
+		if verbose or debug:
+			print("minAdmissibleQ, state", state, "action", action, ":", q)
+
+		return q
+
+	@lru_cache(maxsize=None)
+	def maxAdmissibleV(self, state): # recursive
+		if verbose or debug:
+			print("maxAdmissibleV, state", state, "...")
+
+		actions = self.world.stateToActions(state)
+		qs = [self.maxAdmissibleQ(state, a) for a in actions] # recursion
+		v = max(qs) if maxLambda == 1 else interpolate(min(qs), maxLambda, max(qs))
+
+		if verbose or debug:
+			print("maxAdmissibleV, state", state, ":", v)
+
+		return v
+
+	@lru_cache(maxsize=None)
+	def minAdmissibleV(self, state): # recursive
+		if verbose or debug:
+			print("minAdmissibleV, state", state, "...")
+
+		actions = self.world.stateToActions(state)
+		qs = [self.minAdmissibleQ(state, a) for a in actions] # recursion
+		v = min(qs) if minLambda == 0 else interpolate(min(qs), minLambda, max(qs))
+
+		if verbose or debug:
+			print("minAdmissibleV, state", state, ":", v)
+
+		return v
+
+	# The resulting admissibility interval for states.
+	def admissibility4state(self, state):
+		return self.minAdmissibleV(state), self.maxAdmissibleV(state)
+
+	# The resulting admissibility interval for actions.
+	def admissibility4action(self, state, action):
+		return self.minAdmissibleQ(state, action), self.maxAdmissibleQ(state, action)
+
+
+
+
+	# Some safety metrics do not depend on aspiration and can thus also be computed upfront,
+	# like min/maxAdmissibleQ, min/maxAdmissibleV:
+
+	# Messing potential (maximal entropy (relative to some uninformedStatePrior) 
+	# over trajectories any agent could produce from here (see overleaf for details)):
+	@lru_cache(maxsize=None)
+	def messingPotential_action(self, state, action): # recursive
+		def sample():
+			if state.terminateAfterAction:
+				return 0
+			else:
+				# TODO understand these two lines
+				nextState = self.world.transition(state, action)
+				nextStateScore = self.transitionDistribution(state, action).score(nextState)
+
+				nextMP = self.messingPotential_state(nextState) # recursion
+				priorScore = self["uninformedStatePriorScore"](state) if self["uninformedStatePriorScore"] else 0
+				internalEntropy = self["internalTransitionEntropy"](state, action, nextState) if self["internalTransitionEntropy"] else 0
+				return nextMP + priorScore - nextStateScore + internalEntropy;
+
+		# Note for ANN approximation: messingPotential_action can be positive or negative. 
+		return distribution.infer(sample).E()
```

Comment:
Similarly to above, we can here use the MDPWorldModel's function `` (https://github.com/pik-gane/satisfia/blob/world_model/src/world_model/world_model.py#L64) as follows:
```
def f(nextState, probability):
    if ...:
        return 0
    else:
        ....
        priorScore = self["uninformedStatePriorScore"](nextState) if self["uninformedStatePriorScore"] else 0
        return nextMP + priorScore - np.log(probability) + internalEntropy

return self.world_model.expectation_of_fct_of_probability(state, action, f) 
```
(also note that priorScore should use nextState rather than state. That was a bug in my code which I now fixed).

The reason why the log probability (="score") of the nextState occurs in the function of which we take the expectation is that this is the formula for entropies:  $H(p) = - \sum_x p_x \log(p_x)$

Reply:
In the code expectation_of_fct_of_probability accepts history as a first argument, not just the current state.

Reply:
it should be `[state]`.